demos/signature/EVP_Signature_demo.c

```diff
@@ -0,0 +1,239 @@
+        data =  pub_key_der;
+        data_len = sizeof(pub_key_der);
+    } else {
+        selection =  EVP_PKEY_KEYPAIR;
+        data = priv_key_der;
+        data_len = sizeof(priv_key_der);
+    }
+    dctx = OSSL_DECODER_CTX_new_for_pkey(&pkey, "DER", NULL, "EC",
+                                         selection, libctx, propq);
+    (void)OSSL_DECODER_from_data(dctx, &data, &data_len);
+    OSSL_DECODER_CTX_free(dctx);
+    if (pkey == NULL)
+        fprintf(stderr, "Failed to load %s key.\n", public ? "Public" : "Private");
+    return pkey;
+}
+
+static int demo_sign(OSSL_LIB_CTX *libctx,  const char *sig_name,
+                     size_t *sig_out_len, unsigned char **sig_out_value)
+{
+    int j = 0, result = 0, public = 0;
+    size_t sig_len;
+    unsigned char *sig_value = NULL;
+    const char *propq = NULL;
+    EVP_MD_CTX *sign_context = NULL;
+    EVP_PKEY *priv_key = NULL;
+    
+    /* Get private key */
+    priv_key = get_key(libctx, propq, public);
+    if (priv_key == NULL) {
+        fprintf(stderr, "Get private key failed.\n");
+        goto cleanup;
+    }
+    /*
+     * Make a message signature context to hold temporary state
+     * during signature creation
+     */
+    sign_context = EVP_MD_CTX_new();
+    if (sign_context == NULL) {
+        fprintf(stderr, "EVP_MD_CTX_new failed.\n");
+        goto cleanup;
+    }
+    /*
+     * Initialize the sign context to use the fetched
+     * sign provider.
+     */
+    if (EVP_DigestSignInit_ex(sign_context, NULL, sig_name,
+                              libctx, NULL, priv_key, NULL) != 1) {
+        fprintf(stderr, "EVP_DigestSignInit_ex failed.\n");
+        goto cleanup;
+    }
+    /*
+     * EVP_DigestSignUpdate() can be called several times on the same context
+     * to include additional data.
+     */
+    if (EVP_DigestSignUpdate(sign_context, hamlet_1, strlen(hamlet_1)) != 1) {
+        fprintf(stderr, "EVP_DigestSignUpdate(hamlet_1) failed.\n");
+        goto cleanup;
+    }
+    if (EVP_DigestSignUpdate(sign_context, hamlet_2, strlen(hamlet_2)) != 1) {
+        fprintf(stderr, "EVP_DigestSignUpdate(hamlet_2) failed.\n");
+        goto cleanup;
+    }
+    /* Call EVP_DigestSignFinal to get signature length sig_len */
+    if (EVP_DigestSignFinal(sign_context, NULL, &sig_len) != 1) {
+        fprintf(stderr, "EVP_DigestSignFinal failed.\n");
+        goto cleanup;
+    }
+    if (sig_len <= 0) {
+        fprintf(stderr, "EVP_DigestSignFinal returned invalid signature length.\n");
+        goto cleanup;
+    }
+    sig_value = OPENSSL_malloc(sig_len);
+    if (sig_value == NULL) {
+        fprintf(stderr, "No memory.\n");
+        goto cleanup;
+    }
+    if (EVP_DigestSignFinal(sign_context, sig_value, &sig_len) != 1) {
+        fprintf(stderr, "EVP_DigestSignFinal failed.\n");
+        goto cleanup;
+    }
+    *sig_out_len = sig_len;
+    *sig_out_value = sig_value;
+    fprintf(stdout, "Generating Signature:\n");
+    for (j = 0; j < sig_len; j++)
```

Comment:
You could also use BIO_dump_indent_fp(stdout, sig_value, sig_len, 2);
Remove the j variable in that case..

Reply:
Done, thanks for your comments.